src/prefect/events/related.py

```diff
@@ -91,3 +166,43 @@ async def related_resources_from_run_context(
     ]
 
     return related
+
+
+async def _get_and_cache_related_object(
+    kind: str,
+    role: str,
+    client_method: Callable[[Union[UUID, str]], Awaitable[Optional["ORMBaseModel"]]],
+    obj_id: Union[UUID, str],
+    cache: RelatedResourceCache,
+) -> ResourceCacheEntry:
+    cache_key = f"{kind}.{obj_id}"
+    entry = None
+
+    if cache_key in cache:
+        entry, _ = cache[cache_key]
+    else:
+        obj_ = await client_method(obj_id)
+        entry = {
+            "kind": kind,
+            "object": obj_,
+        }
+
+    cache[cache_key] = (entry, pendulum.now("UTC"))
+
+    # In the case of a worker or agent this cache could be long-lived. To keep
```

Comment:
I tried using `functools.lru_cache` for this and it wasn't happy about anything I tried, especially that I was having it memoize coroutines. I think this simple imlementation of an LRU cache will work for these purposes, but we might consider installing `cachetools` instead.

Reply:
That's pretty slick!  I'm surprised that the `lru_cache` was giving you trouble, though :thinking: